vhost-user-backend/src/handler.rs

```diff
@@ -712,6 +716,41 @@ where
         self.uffd = None;
         Ok(())
     }
+
+    // Sets logging (i.e., bitmap) shared memory space.
+    //
+    // During live migration, the front-end may need to track the modifications the back-end
+    // makes to the memory mapped regions. The front-end should mark the dirty pages in a log.
+    // Once it complies to this logging, it may declare the `VHOST_F_LOG_ALL` vhost feature.
+    //
+    // If the backend has the `VHOST_USER_PROTOCOL_F_LOG_SHMFD` protocol feature it may receive
+    // the `VHOST_USER_SET_LOG_BASE` message. The log memory file descriptor is provided in `file`,
+    // the size and offset of shared memory area are provided in the `VhostUserLog` message.
+    //
+    // See https://qemu-project.gitlab.io/qemu/interop/vhost-user.html#migration.
+    fn set_log_base(&mut self, log: &VhostUserLog, file: File) -> VhostUserResult<()> {
+        let mem = self.atomic_mem.memory();
+
+        // Let's create all bitmaps first before replacing them, in case any of them fails
+        let mut bitmaps = Vec::new();
+        for region in mem.iter() {
+            let logmem = MmapLogReg::from_file(file.as_fd(), log.mmap_offset, log.mmap_size)
+                .map_err(VhostUserError::ReqHandlerError)?;
```

Comment:
Just to avoid an indirection (but not strong opinion on that), but I think is more clear (to me) to see that `MmapLogReg` is `Send` if moved

Reply:
I don’t see the indirection as a problem when every `mark_dirty()` already has to take a read lock on an `RwLock`.

The problem I have is that we `mmap()` the whole bitmap once per region, even though every region only needs access to part of the bitmap, so that seems wasteful. We could shrink the mmapped window appropriately for every region, but that would be complicated, or we just share the bitmap between all regions.

Reply:
That will make more difficult to change vm-memory to reuse their bitmap implementation